provider/azure/environ.go

```diff
@@ -109,6 +109,9 @@ type azureEnviron struct {
 	// modelName is the name of the model.
 	modelName string
 
+	// namespace is used to create the machine and device hostnames.
+	namespace instance.Namespace
```

Comment:
I really like that we're using something that already existed (instance.Namespace). I'm a little concerned that:
1) We have a lot of things called Namespace in the core/ package:
```
cache/modelsummarywatcher.go
33:     Namespace   string
lease/store.go
68:     Namespace string
changestream/subscription.go
28:// Namespace returns the name of the type that the subscription will tied to.
29:func (o SubscriptionOption) Namespace() string {
43:// Namespace returns a SubscriptionOption that will subscribe to the given
45:func Namespace(namespace string, changeMask ChangeType) SubscriptionOption {
instance/namespace.go
18:// Namespace provides a way to generate machine hostanmes with a given prefix.
19:type Namespace interface {
```

This isn't something that you did, but we do need to be careful because we have Namespace that doesn't actually mean the same thing everywhere. (it isn't even always a string)

2) I'm slightly concerned that we'd end up with code that assumes names are deterministic everywhere. I think that was Ian's concern around "what happens when you upgrade to 3.4". I think 3.4 is actually fine, but I can imagine us 3 years in the future to make a 'simplification' in the code where we are doing a lookup, and we assume 'well everything is generated the same way, so I don't need to read the DB for the name, I can just craft the string locally'.

Reply:
Thanks for your review.

1) I see your point, but in this particular case I think it's correctly used because:
  a) It's consistent to how it's used on all other providers (the namespace is always part of the environ struct like this case)
  b) This is consistent to how we use "namespace" in the new dqlite domain in 4.0, as the model db. I find it clear with respect to this convention.

2) I updated the QA steps to include a new controller 3.3->3.4 and model migration (plus new deploys an machine removals) and it works correctly. I do agree though that we'll need to be careful in the future when dealing with lookups.